### PR TITLE
new module for converting the calo fragments directly into caloHits w/o producing caloDigis

### PR DESCRIPTION
**Pull request by gianipez**
- This is the preliminary version of a new module used in the DAQ for converting art Fragments into CaloHits. We will still need the fragment-to-digi converter for debug and in the  DQM.
- added also a utility class to avoid cloning of shared rotines

I would like to push this MR before @miscetti pushed the caloDMAP, so that he would be capable to integrate the channel map also in this module.

### CODE REVIEW

**Comment by kutschke** [comment]
Hi Giani,

  Please look into the failure in ceSimReco.  If you follow the "Logfiles may also be accessed ... " link at the bottom of the posted build/test report, the ceSimReco log file is the top item on the page.   It looks like something in the fcl #include chain within ceSimReco.fcl is still using the old NewCaloClusterFast and not the new CaloClusterOnline.    

   When you change a module label in JobConfig, please check for all uses of the old module label.  The way I do is to do the merge in a scratch git clone and then:

   find . -name \*.fcl -exec grep -H NewCaloClusterFast {} \;

  In this case it finds:

./CaloCluster/fcl/prolog_trigger.fcl
./Analyses/fcl/prolog.fcl

 Note that you also changed the name FastCaloHitMaker.  It is still present in 3 places in:
./CaloReco/fcl/prolog_trigger.fcl

I have no idea if that's a problem or not

Rob

**Comment by gianipez** [comment]
Hi Rob,

I’ll do it. I was not aware that this module was used anywhere already. 

Cheers,

Giani

**Comment by kutschke** [comment]
Hi Giani,

    No worries.   I think it’s Ok to put in a pull request and see what the CI system turns up.

   Rob


On 3/5/21, 11:47 AM, "Gianantonio Pezzullo" <notifications@github.com<mailto:notifications@github.com>> wrote:

Hi Rob,

I’ll do it. I was not aware that this module was used anywhere already.

Cheers,

Giani

**Comment by gianipez** [comment]
please Jenkins try a build

**Comment by kutschke** [comment]
@FNALbuild run build tests

File: RecoDataProducts/inc/CaloHit.hh
Comment (kutschke) [inline RecoDataProducts/inc/CaloHit.hh:47]: Is this change just for aesthetic reasons?

I think that root's automatic schema evolution will manage this correctly but it may mean that reading old files with new code will be slower.  Is that an issue that we need to think about?  If that's only a small transient problem then this is a good time for the change.

Comment (gianipez) [reply RecoDataProducts/inc/CaloHit.hh:47]: I don't think we will have an issue with older datasets as there were other breaking changes, no? I changed that because in the online reco we don't fill the *Err_ parameters and @bechenard suggested me to make this change to optimize the memory allocation.

File: DAQ/src/CaloRecoFromFragments_module.cc
Comment (kutschke) [inline DAQ/src/CaloRecoFromFragments_module.cc:256]: Why do you need this cast?  I don't think that it does anything?  

In cases where you do need a cast, please use c++ style casts.

File: RecoDataProducts/inc/CaloHit.hh
Comment (kutschke) [reply RecoDataProducts/inc/CaloHit.hh:47]: Yes, we are making other breaking changes.  That's why I said this is a good time for a change.   I will mark this resolved but I have a  follow up offline.

File: DAQ/src/CaloRecoFromFragments_module.cc
Comment (gianipez) [reply DAQ/src/CaloRecoFromFragments_module.cc:256]: done.

Comment (kutschke) [reply DAQ/src/CaloRecoFromFragments_module.cc:256]: There are many others of these in this file.  Please fix them all and I will merge right after.

Comment (gianipez) [reply DAQ/src/CaloRecoFromFragments_module.cc:256]: those I left are needed because they are uint8/16 and when I was printing them with cout was not readable.

Comment (kutschke) [reply DAQ/src/CaloRecoFromFragments_module.cc:256]: OK.  Merging now.

**Review by kutschke — APPROVED** [review]

File: DAQ/src/CaloHitsFromFragments_module.cc
Comment (bonventre) [inline DAQ/src/CaloHitsFromFragments_module.cc:209]: Can this part not go in the utility so its not duplicated from StrawAndCaloDigisFromFragments?

Comment (gianipez) [reply DAQ/src/CaloHitsFromFragments_module.cc:209]: The operations inside the loop are different. Both modules are looping over the same object, but they produce different outputs that need different processing.

Comment (kutschke) [reply DAQ/src/CaloHitsFromFragments_module.cc:209]: I see that I jumped the gun on the merge.  Sorry about that.  If another small PR is warranted I will watch for it.

I am not sure that I understand.  I think that Richie is talking about the call to the print utility at line 206 but Giani is talking about code in the loop that starts at 209?  Can you straighten me out?

Comment (bonventre) [reply DAQ/src/CaloHitsFromFragments_module.cc:209]: We're talking about the loop - we'll just have to be careful to keep these both in sync

**Pull request merged**
